File: HubPrediction/Forecast.swift
import Foundation

enum Forecast {
    static let maxSlope = 2.2
    static let emaKeep = 0.82
    static let emaNew = 0.18
    /// Always reserve this much future so the next-15m dash sits in-frame.
    static let futurePadMs = 16.0 * HubMs.minute

    static func clamp(_ n: Double, _ lo: Double, _ hi: Double) -> Double {
        min(hi, max(lo, n))
    }

    static func slopeFromPoints(_ points: [Point]?, now: Double) -> Double {
        let list = points ?? []
        let from = now - 6.0 * HubMs.minute
        let slice = list.filter { $0.t.isFinite && $0.px.isFinite && $0.t >= from && $0.t <= now + 1.5 * HubMs.second }
        guard slice.count >= 2, let a = slice.first, let b = slice.last else { return 0 }
        let mins = (b.t - a.t) / HubMs.minute
        if mins < 0.45 { return 0 }
        return clamp((b.px - a.px) / mins, -maxSlope, maxSlope)
    }

    static func emaSlope(prev: Double?, raw: Double) -> Double {
        guard let prev, prev.isFinite else { return raw }
        return prev * emaKeep + raw * emaNew
    }

    static func forwardRay(now: Double, closeAt: Double, live: Double, slopePerMin: Double, lean: DeskSide) -> [Point] {
        guard live.isFinite, now.isFinite else { return [] }
        var slope = clamp(slopePerMin, -maxSlope, maxSlope)
        if abs(slope) < 0.18 && lean != .sit {
            slope += lean == .up ? 0.12 : -0.12
        }
        let close = (closeAt.isFinite && closeAt > now) ? closeAt : now + 15.0 * HubMs.minute
        let lastT = max(now + futurePadMs, close + 15.0 * HubMs.minute)
        var times: [Double] = []
        var t = now
        while t < lastT - 1 {
            times.append(t)
            t += HubMs.minute
        }
        times.append(lastT)
        return times.map { Point(t: $0, px: live + slope * (($0 - now) / HubMs.minute)) }
    }

    /// History on the left, at least 16m of future on the right so the dash is on-canvas.
    static func chartWindow(now: Double, zoomMinutes: Double, pan: Double, lastForecastT: Double?) -> (start: Double, end: Double) {
        let cursor = now + pan
        let start = cursor - zoomMinutes * HubMs.minute
        let pad = max(futurePadMs, (lastForecastT ?? now) - now)
        return (start, cursor + pad)
    }

    static func rebasePrior(_ prior: [Point]?, live: Double) -> [Point] {
        let list = (prior ?? []).filter { $0.px.isFinite && $0.t.isFinite }
        guard let last = list.last?.px, last.isFinite, live.isFinite else { return [] }
        let shift = live - last
        return list.map { Point(t: $0.t, px: $0.px + shift) }
    }

    static func yDomain(live: Double, values: [Double]) -> (Double, Double) {
        let base = live.isFinite ? live : 0
        let clamped = values.filter { $0.isFinite && abs($0 - base) < 250 }
        let lo = min(base - 20.0, clamped.min() ?? base - 20.0)
        let hi = max(base + 20.0, clamped.max() ?? base + 20.0)
        if !lo.isFinite || !hi.isFinite { return (base - 20.0, base + 20.0) }
        return (lo, hi)
    }
}

File: hub-prediction/src/lib/forecast.ts
import type { DeskSide, Point } from './types'

export const MAX_SLOPE = 2.2
export const EMA_KEEP = 0.82
export const EMA_NEW = 0.18

function clamp(n: number, lo: number, hi: number) {
  return Math.max(lo, Math.min(hi, n))
}

export function slopeFromPoints(points: Point[] | undefined, now: number) {
  const list = Array.isArray(points) ? points : []
  const from = now - 6 * 60_000
  const slice = list.filter((p) => p && Number.isFinite(p.t) && Number.isFinite(p.px) && p.t >= from && p.t <= now + 1500)
  if (slice.length < 2) return 0
  const a = slice[0]
  const b = slice[slice.length - 1]
  const mins = (b.t - a.t) / 60_000
  if (mins < 0.45) return 0
  return clamp((b.px - a.px) / mins, -MAX_SLOPE, MAX_SLOPE)
}

export function emaSlope(prev: number | null, raw: number) {
  if (prev == null || !Number.isFinite(prev)) return raw
  return prev * EMA_KEEP + raw * EMA_NEW
}

export function forwardRay(opts: {
  now: number
  closeAt: number
  live: number
  slopePerMin: number
  lean: DeskSide
}): Point[] {
  const live = opts.live
  const now = opts.now
  if (!Number.isFinite(live) || !Number.isFinite(now)) return []

  let slope = clamp(opts.slopePerMin, -MAX_SLOPE, MAX_SLOPE)
  if (Math.abs(slope) < 0.18 && opts.lean !== 'sit') {
    slope += opts.lean === 'up' ? 0.12 : -0.12
  }

  const closeAt = Number.isFinite(opts.closeAt) && opts.closeAt > now ? opts.closeAt : now + 15 * 60_000
  const lastT = Math.max(now + 16 * 60_000, closeAt + 15 * 60_000)
  const pxAt = (t: number) => live + slope * ((t - now) / 60_000)
  const times: number[] = []
  for (let t = now; t < lastT - 1; t += 60_000) times.push(t)
  times.push(lastT)
  return times.map((t) => ({ t, px: pxAt(t) }))
}

export function rebasePrior(prior: Point[] | undefined, live: number): Point[] {
  const list = Array.isArray(prior) ? prior.filter((p) => p && Number.isFinite(p.px) && Number.isFinite(p.t)) : []
  if (!list.length || !Number.isFinite(live)) return []
  const last = list[list.length - 1]?.px
  if (!Number.isFinite(last)) return []
  const shift = live - last
  return list.map((p) => ({ t: p.t, px: p.px + shift }))
}

export function yDomain(live: number, values: number[]): [number, number] {
  const base = Number.isFinite(live) ? live : 0
  const clamped = (values ?? []).filter((v) => Number.isFinite(v) && Math.abs(v - base) < 250)
  const lo = Math.min(base - 20, ...clamped)
  const hi = Math.max(base + 20, ...clamped)
  if (!Number.isFinite(lo) || !Number.isFinite(hi)) return [base - 20, base + 20]
  return [lo, hi]
}

export function maxStep(points: Point[]) {
  let max = 0
  const list = Array.isArray(points) ? points : []
  for (let i = 1; i < list.length; i++) {
    max = Math.max(max, Math.abs(list[i].px - list[i - 1].px))
  }
  return max
}

/** Grok Build / buildDashboard theory — fade + last-week shape, beats naive slope. */
export const THEORY_FADE_MINS = 90
export const THEORY_CLAMP = 10

export function slotTheory(opts: {
  t: number
  nowSlot: number
  lookNow: number
  live: number
  slope: number
  lastWeek: number | null
  lastWeekNow: number
  actual: number | null
}): number | null {
  const mins = (opts.t - opts.lookNow) / 60_000
  const fade = Math.max(0, 1 - Math.max(0, mins) / THEORY_FADE_MINS)
  const shape = opts.lastWeek != null ? opts.lastWeek - opts.lastWeekNow : 0
  const slope = clamp(opts.slope, -MAX_SLOPE, MAX_SLOPE)
  if (opts.t >= opts.nowSlot) {
    return opts.live + slope * Math.max(0, mins) * fade + shape
  }
  if (opts.actual != null) {
    const raw = opts.lastWeek != null ? opts.live + shape : opts.actual
    return Math.max(opts.actual - THEORY_CLAMP, Math.min(opts.actual + THEORY_CLAMP, raw))
  }
  return opts.lastWeek != null ? opts.live + shape : null
}

export function slotPreview(opts: {
  t: number
  nowSlot: number
  lookNow: number
  live: number
  slope: number
  actual: number | null
}): number | null {
  if (opts.actual != null) return null
  if (opts.t < opts.nowSlot) return null
  const mins = (opts.t - opts.lookNow) / 60_000
  return opts.live + clamp(opts.slope, -MAX_SLOPE, MAX_SLOPE) * Math.max(0, mins)
}

export function vsOpen(lastWeek: number | null, dayOpen: number | null) {
  if (lastWeek == null || dayOpen == null) return null
  if (!Number.isFinite(lastWeek) || !Number.isFinite(dayOpen)) return null
  return lastWeek - dayOpen
}

export function theoryBeatsNaive(actual: number, theory: number, preview: number) {
  return Math.abs(actual - theory) < Math.abs(actual - preview)
}
